### PR TITLE
fix: Enforce NonAdminBackup spec field values

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Why the changes were made

Resolves https://github.com/migtools/oadp-non-admin/discussions/33

Related to: https://github.com/migtools/oadp-non-admin/issues/37

~Blocked by: https://github.com/openshift/oadp-operator/pull/1584~

This PR adds fetching of OADP's DPA during NAC set up, adding DPA `spec.nonAdmin.enforceBackupSpec` to `NonAdminBackupReconciler` and adds validation for NonAdminBackup spec respecting enforcement.

> **Note:** admin user can not enforce falsy values.

## How to test the changes made

Read [design](https://github.com/mateusoliveira43/oadp-non-admin/blob/fix/enforce-backup-spec-field-values/docs/design/admin_control_over_spec.md).

Follow [install-from-source](https://github.com/migtools/oadp-non-admin/blob/master/docs/CONTRIBUTING.md#install-from-source) testing documentation, pointing this branch and my OADP fork branch [`fix/enforce-non-admin-bckup-spec-field-values`](https://github.com/mateusoliveira43/oadp-operator/tree/fix/enforce-non-admin-bckup-spec-field-values)

### CODE REVIEW

File: internal/common/function/function.go
Comment (mateusoliveira43) [inline internal/common/function/function.go:1]: TODO missing design

File: go.mod [hunk 1]
```diff
@@ -70,10 +74,11 @@ require (
...
 	k8s.io/kube-openapi v0.0.0-20231010175941-2dd684a91f00 // indirect
-	k8s.io/utils v0.0.0-20230726121419-3b25d923346b // indirect
 	sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.4.1 // indirect
 	sigs.k8s.io/yaml v1.4.0 // indirect
 )
 
 replace github.com/vmware-tanzu/velero => github.com/openshift/velero v0.10.2-0.20240919150610-92244630d90b
+
+replace github.com/openshift/oadp-operator => github.com/mateusoliveira43/oadp-operator v0.0.0-20241108184219-7913bfce1719
```
Comment (mateusoliveira43) [inline go.mod:84]: TODO remove this prior to merge

File: docs/design/admin_control_over_spec.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Design proposal template `Admin users control over non admin objects' spec`
```
Comment (shubham-pampattiwar) [inline docs/design/admin_control_over_spec.md:1]: Remove template placeholder - "Design Proposal template"

File: docs/design/admin_control_over_spec.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+## Abstract
+
+Non Admin Controller (NAC) restricts the usage of OADP operator with NonAdminBackups and NonAdminRestores.
+Admin users may want to further restrict this by restricting NonAdminBackup/NonAdminRestore spec fields values.
+
+## Background
+
+Non Admin Controller (NAC) adds the ability to admin users restrict the use of OADP operator for non admin users, by only allowing them to create backup/restores from their namespaces with NonAdminBackups/NonAdminRestores.
+Admin users may want to further restrict non admin users operations, like forcing a specific time to live (TTL) for NonAdminBackups associated Velero Backups.
+This design enables admin users to set custom default values for NonAdminBackup/NonAdminRestore spec fields, which can not be overridden.
```
Comment (shubham-pampattiwar) [inline docs/design/admin_control_over_spec.md:12]: can not be overridden by non-admin users...

File: docs/design/admin_control_over_spec.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+## Goals
+
+Enable admin users to
+- set custom default values for NonAdminBackup spec.backupSpec fields, which can not be overridden
+- TODO restore
+
+## Non Goals
+
+- Show NonAdminBackup spec.backupSpec fields/TODO NonAdminRestore custom default values to non admin users
+- Do not allow non admin users to create NonAdminBackup/NonAdminRestore with overridden defaults (but they will not validated by NAC, with an error explaining why in the object statuses and in NAC logs)
```
Comment (shubham-pampattiwar) [inline docs/design/admin_control_over_spec.md:23]: can you please simplify this non-goal here ?

File: docs/design/admin_control_over_spec.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+## High-Level Design
+
+A field will be added to OADP DPA object. With it, admin users will be able to select which NonAdminBackup `spec.backupSpec` fields have custom default (and enforced) values. NAC will respect the set values. If a NonAdminBackup is created with fields overriding any enforced values, it will fail validation prior to creating an associated Velero Backup. If admin user changes any enforced field value, NAC Pod is recreated to always be up to date with admin user enforcements.
+
+TODO restore
+
+
+## Detailed Design
+
+Field `spec.nonAdmin.enforceBackupSpec`, of the same type as the Velero Backup Spec, will be added to OADP DPA object.
```
Comment (shubham-pampattiwar) [inline docs/design/admin_control_over_spec.md:36]: Please give an example of DPA yaml file here. May go over a use-case/scenario of how the workflow would be ?
- Sample DPA configured by admin
- Sample NAB case accepting value from Non-Admin user and al respecting some defaults from Admin enforced values

File: docs/design/admin_control_over_spec.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+- Show NonAdminBackup spec.backupSpec fields/TODO NonAdminRestore custom default values to non admin users
+- Do not allow non admin users to create NonAdminBackup/NonAdminRestore with overridden defaults (but they will not validated by NAC, with an error explaining why in the object statuses and in NAC logs)
+- Allow admin users to set second level defaults (for example, NonAdminBackup `spec.backupSpec.labelSelector` can have a custom default value, but not just `spec.backupSpec.labelSelector.matchLabels`)
+
+
+## High-Level Design
+
+A field will be added to OADP DPA object. With it, admin users will be able to select which NonAdminBackup `spec.backupSpec` fields have custom default (and enforced) values. NAC will respect the set values. If a NonAdminBackup is created with fields overriding any enforced values, it will fail validation prior to creating an associated Velero Backup. If admin user changes any enforced field value, NAC Pod is recreated to always be up to date with admin user enforcements.
+
```
Comment (shubham-pampattiwar) [inline docs/design/admin_control_over_spec.md:30]: We should document the fact that if NAC is restarted in an on-going NAB operation, reconcile progress might get lost.

File: docs/design/admin_control_over_spec.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+## Security Considerations
+
+No security considerations.
+
+## Compatibility
+
+No compatibility issues.
+
+## Implementation
+
```
Comment (shubham-pampattiwar) [inline docs/design/admin_control_over_spec.md:63]: Lets add some text here explaining the logic that's getting implemented. For instance you could say:
- Add Enforced Backup Spec to NAB reconciler struct
- Modify ValidateSpec function to use EnforcedBackup spec and apply that to users' NAB request

**Review by shubham-pampattiwar — CHANGES_REQUESTED** [review]

File: cmd/main.go [hunk 8]
```diff
@@ -104,7 +110,32 @@ func main() {
...
+	dpaClientScheme := runtime.NewScheme()
+	utilruntime.Must(v1alpha1.AddToScheme(dpaClientScheme))
+	dpaClient, err := client.New(restConfig, client.Options{
+		Scheme: dpaClientScheme,
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to create Kubernetes client")
+		os.Exit(1)
+	}
+	// TODO we could pass DPA name as env var and do a get call directly. Better?
```
Comment (mateusoliveira43) [inline cmd/main.go:124]: TODO

File: internal/common/function/function_test.go [hunk 9]
```diff
@@ -122,6 +127,254 @@ func TestValidateBackupSpec(t *testing.T) {
...
+					"openshift/banana": "false",
+					constant.OadpLabel: constant.TrueString,
+				},
+			},
+		},
+		{
+			name: "IncludedNamespaces",
+		},
+		{
+			// should not be enforced?
```
Comment (mateusoliveira43) [inline internal/common/function/function_test.go:158]: TODO

File: cmd/main.go [hunk 10]
```diff
@@ -104,7 +110,32 @@ func main() {
 		os.Exit(1)
 	}
 
-	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+	restConfig := ctrl.GetConfigOrDie()
+
+	dpaClientScheme := runtime.NewScheme()
```
Comment (shubham-pampattiwar) [inline cmd/main.go:115]: Any reason we cannot just add the OADP DPA scheme to reconciler client and fetch the enforced Backup spec at the start of NAB reconciler ? before stepping into the reconcile step functions ?

Comment (mateusoliveira43) [reply cmd/main.go:115]: that would do a get call on every reconcile, no?

preferred to do here to do just one call and do not add oadp scheme to reconcile client

Comment (mpryc) [reply cmd/main.go:115]: I think @shubham-pampattiwar  was thinking of adding it to the:

```
func init() {
	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
	utilruntime.Must(nacv1alpha1.AddToScheme(scheme))
	utilruntime.Must(velerov1.AddToScheme(scheme))
```

And then using as `Scheme` within reconcile. This way it behaves same as others. The get on that can be limited by predicate ?

Comment (shubham-pampattiwar) [reply cmd/main.go:115]: I was thinking of just adding scheme here to the controller client, and then in NAB reconcile loop we could have a task func called `refreshEnforcedBackupSpec`. In this function you only update the in-memory reconciler struct if there was a change in enforcedspec via DPA. To limit the Get calls to DPA we could use the same annotation we used for restarted nac pod, `enforced-backup-spec` to determine if refresh of  enforcedspec is needed or not. Thoughts on this ?

**Review by shubham-pampattiwar — DISMISSED** [review]
Changes worked fine for me:
```
spec:
  backupSpec:
    includedNamespaces:
    - nacuser
    snapshotVolumes: true
status:
  conditions:
  - lastTransitionTime: "2024-11-20T04:42:43Z"
    message: NonAdminBackup spec.backupSpec.snapshotVolumes field value is enforced
      by admin user, can not override it
    reason: InvalidBackupSpec
    status: "False"
    type: Accepted
  phase: BackingOff
```

**Review by mpryc — CHANGES_REQUESTED** [review]
1. Crash when no dpa.Spec.NonAdmin.EnforceBackupSpec is specified:

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x15d50bb]

goroutine 1 [running]:
main.getEnforcedSpec(0xc00042a908, {0xc00004c0f0, 0xd})
	cmd/main.go:195 +0x1db
main.main()
	cmd/main.go:116 +0x4fa
```

Fix:
```diff
diff --git a/cmd/main.go b/cmd/main.go
index 0d32e48..b8b50be 100644
--- a/cmd/main.go
+++ b/cmd/main.go
@@ -30,6 +30,7 @@ import (
        "k8s.io/apimachinery/pkg/runtime"
        utilruntime "k8s.io/apimachinery/pkg/util/runtime"
        clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+
        // Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
        // to ensure that exec-entrypoint and run can make use of them.
        _ "k8s.io/client-go/plugin/pkg/client/auth"
@@ -192,7 +193,9 @@ func getEnforcedSpec(restConfig *rest.Config, oadpNamespace string) (*velerov1.B
        enforcedBackupSpec := &velerov1.BackupSpec{}
        for _, dpa := range dpaList.Items {
                if dpa.Namespace == oadpNamespace {
-                       enforcedBackupSpec = dpa.Spec.NonAdmin.EnforceBackupSpec
+                       if dpa.Spec.NonAdmin != nil && dpa.Spec.NonAdmin.EnforceBackupSpec != nil {
+                               enforcedBackupSpec = dpa.Spec.NonAdmin.EnforceBackupSpec
+                       }
                }
        }
        return enforcedBackupSpec, nil
```

1. Create the NonAdminBackup with different then enforced flag
Result: Proper BackingOff state and condition

First enforce failure:

```yaml
status:
  conditions:
    - lastTransitionTime: '2024-11-21T11:19:58Z'
      message: 'NonAdminBackup spec.backupSpec.snapshotVolumes field value is enforced by admin user, can not override it'
      reason: InvalidBackupSpec
      status: 'False'
      type: Accepted
  phase: BackingOff
```

2. Patch the NonAdminBackup with proper enforced flag
Result: Proper Created state and condition

3. Patch again the NonAdminBackup with different then enforced flag
Result: roper BackingOff state and condition, veleroBackup reference is also proper as it was from the previous patch.

```yaml
status:
  conditions:
    - lastTransitionTime: '2024-11-21T11:23:03Z'
      message: 'NonAdminBackup spec.backupSpec.snapshotVolumes field value is enforced by admin user, can not override it'
      reason: InvalidBackupSpec
      status: 'False'
      type: Accepted
    - lastTransitionTime: '2024-11-21T11:21:05Z'
      message: Created Velero Backup object
      reason: BackupScheduled
      status: 'True'
      type: Queued
  phase: BackingOff
  veleroBackup:
    nacuuid: default-exampleabc-8a10e3ec-0610-429a-86cf-421d5da151b2
    name: default-exampleabc-8a10e3ec-0610-429a-86cf-421d5da151b2
    namespace: openshift-adp
    status: {}
```

**Comment by mpryc** [comment]
@shubham-pampattiwar there was also small lint issue while running `make lint` on that PR, should be fixed now:

```
internal/common/function/function.go:85:21: cannot range over enforcedSpec.NumField() (value of type int) (typecheck)
	for index := range enforcedSpec.NumField() {
	                   ^
internal/common/function/function_test.go:363:22: cannot range over backupSpec.NumField() (value of type int) (typecheck)
		for index := range backupSpec.NumField() {
		                   ^
internal/controller/nonadminbackup_controller.go:610:22: cannot range over enforcedSpec.NumField() (value of type int) (typecheck)
		for index := range enforcedSpec.NumField() {
		                   ^
make: *** [Makefile:75: lint] Error 1



cmd/main.go:33: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/migtools/oadp-non-admin) (gci)

```

**Review by shubham-pampattiwar — APPROVED** [review]

**Comment by mpryc** [comment]
override snyk as it's not problematic atm

**Pull request merged**